linera-service/src/benchmark.rs

```diff
@@ -0,0 +1,266 @@
+    let bytecode_id = publisher.publish_bytecode(contract, service, None).await?;
+
+    struct BenchmarkContext {
+        application_id: ApplicationId<FungibleTokenAbi>,
+        owner: Owner,
+        default_chain: ChainId,
+    }
+
+    // Create the fungible applications
+    let apps = join_all(clients.iter().zip(services).enumerate().map(
+        |(i, (client, node_service))| async move {
+            let owner = client.get_owner().unwrap();
+            let default_chain = client.default_chain().unwrap();
+            let initial_state = InitialState {
+                accounts: BTreeMap::from([(
+                    AccountOwner::User(client.get_owner().unwrap()),
+                    Amount::from_tokens(n_transactions as u128),
+                )]),
+            };
+            let parameters = Parameters::new(format!("FUN{}", i).leak());
+            let application_id = node_service
+                .create_application::<FungibleTokenAbi>(
+                    &default_chain,
+                    &bytecode_id,
+                    &parameters,
+                    &initial_state,
+                    &[],
+                )
+                .await
+                .unwrap();
+            let context = BenchmarkContext {
+                application_id,
+                owner,
+                default_chain,
+            };
+            let app = FungibleApp(
+                node_service
+                    .make_application(&context.default_chain, &context.application_id)
+                    .await
+                    .unwrap(),
+            );
+            (app, context, node_service)
+        },
+    ))
+    .await;
+
+    // create transaction futures
+    let mut expected_balances = vec![vec![Amount::ZERO; apps.len()]; apps.len()];
+    let mut rng = rand::rngs::SmallRng::seed_from_u64(seed);
+
+    let transaction_futures = (0..n_transactions).flat_map(|transaction_i| {
+        apps.iter()
+            .enumerate()
+            .map(|(sender_i, (sender_app, sender_context, _))| {
+                let receiver_i = if uniform {
+                    (transaction_i + sender_i + 1) % apps.len()
+                } else {
+                    rng.gen_range(0..apps.len())
+                };
+                let (_, receiver_context, _) = &apps[receiver_i];
+                expected_balances[receiver_i][sender_i]
+                    .try_add_assign(Amount::ONE)
+                    .unwrap();
+                sender_app.transfer(
+                    AccountOwner::User(sender_context.owner),
+                    Amount::ONE,
+                    Account {
+                        chain_id: receiver_context.default_chain,
+                        owner: AccountOwner::User(receiver_context.owner),
+                    },
+                )
+            })
+            .collect::<Vec<_>>()
+    });
+
+    println!("Making {} transactions", n_users * n_transactions);
```

Comment:
This could change in the future but currently, we consistently use `stdout` only for machine-readable outputs (likely newline-separated JSON), and use `tracing` outputting to `stderr` for human-readable outputs. (See CONTRIBUTING.md / "Command-line tools and services")

Reply:
Right, we should probably actually use `info!` for most of these, and then also output the benchmark results as JSON to `stdout`?